test/functional/wallet_basic.py

```diff
@@ -504,15 +504,58 @@ def run_test(self):
                                 {"address": address_to_import},
                                 {"spendable": False})
 
-            # 5. Import private key of the previously imported address on node1
+            # 5. Remove the address from node1
+            self.nodes[1].removeaddress(address_to_import, False, False)
+
+            # 6. Validate tx is still in wallet
+            self.nodes[1].gettransaction(addr_txid)
+
+            # 7. Re-import address then purge transactions
+            self.nodes[1].importaddress(address_to_import)
+            self.nodes[1].removeaddress(address_to_import, False, True)
+
+            # 8. Validate transaction is removed
+            assert_raises_rpc_error(-5, "Invalid or non-wallet transaction id", self.nodes[1].gettransaction, addr_txid)
+
+            # 9. Validate that the address is no longer in the wallet
+            assert not self.nodes[1].getaddressinfo(address_to_import)["iswatchonly"]
+
+            # 10. Attempt to remove address again
+            assert_raises_rpc_error(-4, "The wallet does not contain this address or script", self.nodes[1].removeaddress, address_to_import)
+
+            # 11. Re-import the address
+            self.nodes[1].importaddress(address_to_import)
+
+            # 12. Import private key of the previously imported address on node1
             priv_key = self.nodes[2].dumpprivkey(address_to_import)
             self.nodes[1].importprivkey(priv_key)
 
-            # 6. Check that the unspents are now spendable on node1
+            # 13. Check that the unspents are now spendable on node1
             assert_array_result(self.nodes[1].listunspent(),
                                 {"address": address_to_import},
                                 {"spendable": True})
 
+            # 14. Attempt to remove address after owning private key
+            assert_raises_rpc_error(-4, "The wallet contains the private key for this address", self.nodes[1].removeaddress, address_to_import)
+
+            # Check for removeaddress affecting other addresses
+            # Send some coins to generate new UTXO
+            address_to_import2 = self.nodes[2].getnewaddress()
+            txid = self.nodes[0].sendtoaddress(address_to_import2, 1)
+            self.nodes[0].generate(1)
+            self.sync_all(self.nodes[0:3])
+
+            # Import the new address from node2 to node1
+            self.nodes[1].importaddress(address_to_import2)
+            assert self.nodes[1].getaddressinfo(address_to_import2)["iswatchonly"]
+
+            # Remove the address from node1
+            self.nodes[1].removeaddress(address_to_import2)
+
+            # Check the address was removed and other address is unaffected
+            assert not self.nodes[1].getaddressinfo(address_to_import2)["iswatchonly"]
+            assert self.nodes[1].getaddressinfo(address_to_import)["ismine"]
+
```

Comment:
It would be good to test the `p2sh` flag as well, all these examples only touch addresses not scripts

Reply:
Added.